lib/libvfio-user.c

```diff
@@ -546,14 +547,46 @@ handle_dma_unmap(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg,
     }
 
     char rstr[1024];
-
-    snprintf(rstr, sizeof(rstr), "[%#lx, %#lx) offset=%#lx "
-            "prot=%#x flags=%#x", region->addr, region->addr + region->size,
-            region->offset, region->prot, region->flags);
+    snprintf(rstr, sizeof(rstr), "[%#lx, %#lx) offset=%#lx prot=%#x flags=%#x",
+             region->addr, region->addr + region->size,
+             region->offset, region->prot, region->flags);
 
     vfu_log(vfu_ctx, LOG_DEBUG, "removing DMA region %s", rstr);
 
-    if (region->flags == 0) {
+    if (region->flags == VFIO_DMA_UNMAP_FLAG_GET_DIRTY_BITMAP) {
+        if (msg->in_size != sizeof(struct vfio_user_dma_region) + sizeof(struct vfio_user_bitmap)) {
+            vfu_log(vfu_ctx, LOG_ERR, "bad message size %#lx", msg->in_size);
+            return ERROR_INT(EINVAL);
+        }
+        /*
+         * TODO this could be a separate function, but the implementation is
+         * temprorary anyway since we're moving dirty page tracking out of
+         * the DMA controller.
+         */
+        ret = dma_controller_dirty_page_get(vfu_ctx->dma,
+                                            (vfu_dma_addr_t)region->addr,
+                                            region->size,
+                                            region->bitmap->pgsize,
+                                            region->bitmap->size,
+                                            &bitmap);
+        if (ret < 0) {
+            vfu_log(vfu_ctx, LOG_ERR, "failed to get dirty page bitmap: %m");
+            return -1;
+        }
+        msg->out_iovecs = malloc(sizeof(struct iovec));
+        if (msg->out_iovecs == NULL) {
+            return -1;
+        }
+        msg->out_iovecs->iov_base = malloc(region->bitmap->size);
+        if (msg->out_iovecs == NULL) {
+            free(msg->out_iovecs);
+            msg->out_iovecs = NULL;
+            return ERROR_INT(ENOMEM);
+        }
+        memcpy(msg->out_iovecs->iov_base, bitmap, region->bitmap->size);
```

Comment:
I don't get why we're doing a copy here rather than just populating your allocation?

Reply:
I'm not sure I understand your question, do you mean why not doing `msg->out_iovecs->iov_base = bitmap`? If so, the reason is that `dma_controller_remove_region` is supposed to free `bitmap`, but I just discovered that it doesn't, so that's a memleak. I suppose we can do:
```
msg->out_iovecs->iov_base = region->bitmap;
region->bitmap = NULL;
...
Right?

Reply:
We don't free iovecs in freemsg() either, but we can add that facility. It's certainly better than two allocations?

Reply:
Actually you don't need to use out_iovecs at all since it's just one entry. Why not out_data?

Reply:
good point